lib/src/stream_queue.dart

```diff
@@ -543,28 +600,47 @@ class StreamQueueTransaction<T> {
   /// The parent queue on which this transaction is active.
   final StreamQueue<T> _parent;
 
-  /// The splitter that produces copies of the parent queue's stream.
-  final StreamSplitter<T> _splitter;
-
   /// Queues created using [newQueue].
-  final _queues = <StreamQueue>{};
+  final _queues = <StreamQueue<T>>{};
+
+  /// The value -1 until committed or rejected, then then number of consumed
+  /// results + 1 for committed and 0 for rejected).
+  int _state = -1;
 
-  /// Whether [commit] has been called.
-  var _committed = false;
+  bool get _active => _state < 0;
+  bool get _committed => _state > 0;
+  bool get _rejected => _state == 0;
 
-  /// Whether [reject] has been called.
-  var _rejected = false;
+  int get _consumedEvents {
+    assert(_state >= 0);
+    return _state == 0 ? 0 : _state - 1;
+  }
+
+  void _update(Queue<Result<T>> events, bool isDone) {
+    if (!_active) return;
+    for (var queue in _queues) {
+      var start = queue._eventsReceived;
+      var end = events.length;
+      if (start < end) {
+        queue._addResults(events, start, end, isDone);
+      } else if (isDone) {
+        queue._close();
+      }
+    }
+  }
 
-  StreamQueueTransaction._(this._parent, Stream<T> source)
-      : _splitter = StreamSplitter(source);
+  StreamQueueTransaction._(this._parent);
 
   /// Creates a new copy of the parent queue.
   ///
   /// This copy starts at the parent queue's position when
   /// [StreamQueue.startTransaction] was called. Its position can be committed
   /// to the parent queue using [commit].
   StreamQueue<T> newQueue() {
-    var queue = StreamQueue(_splitter.split());
+    if (!_active) {
+      return StreamQueue._().._flags = StreamQueue._flagDone;
+    }
+    var queue = StreamQueue<T>._init(_parent._eventQueue, _parent._isDone);
```

Comment:
An alternative would be to make the transaction `StreamQueue` pull its parent's events one at a time, when it has a request enqueued. That way it won't copy events that it hasn't asked for, and then we can cut the link when cancelling.
I think that would bring us close to the current behavior. It's more complicated, but might avoid copying events that aren't needed yet. 

Probably needs an asynchronous delay as well, so it'll have to start a microtask to request events.

Reply:
Pulling does sound like a good approach at a high level, but I haven't thought through any details.

> Probably needs an asynchronous delay as well, so it'll have to start a microtask to request events.

Why is that? When it pulls an even it should get it asynchronously already right?

Reply:
If adding a request would start pulling events from the parent queue synchronously, until the request is completed, then the request will be able to complete synchronously.
That might not be a problem, if it just completes an asynchronous completer, nobody will know.
I'm just not absolutely sure that everything that happens is asynchronous.